sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -0,0 +1,367 @@
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, ReturnAnswer}
+import org.apache.spark.sql.catalyst.rules.{Rule, RuleExecutor}
+import org.apache.spark.sql.execution._
+import org.apache.spark.sql.execution.exchange._
+import org.apache.spark.sql.execution.ui.SparkListenerSQLAdaptiveExecutionUpdate
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.ThreadUtils
+
+/**
+ * A root node to execute the query plan adaptively. It splits the query plan into independent
+ * stages and executes them in order according to their dependencies. The query stage
+ * materializes its output at the end. When one stage completes, the data statistics of the
+ * materialized output will be used to optimize the remainder of the query.
+ *
+ * To create query stages, we traverse the query tree bottom up. When we hit an exchange node,
+ * and if all the child query stages of this exchange node are materialized, we create a new
+ * query stage for this exchange node. The new stage is then materialized asynchronously once it
+ * is created.
+ *
+ * When one query stage finishes materialization, the rest query is re-optimized and planned based
+ * on the latest statistics provided by all materialized stages. Then we traverse the query plan
+ * again and create more stages if possible. After all stages have been materialized, we execute
+ * the rest of the plan.
+ */
+case class AdaptiveSparkPlanExec(
+    initialPlan: SparkPlan,
+    session: SparkSession,
+    subqueryMap: Map[Long, ExecSubqueryExpression],
+    stageCache: TrieMap[StructType, mutable.Buffer[(Exchange, QueryStageExec)]])
+  extends LeafExecNode {
+
+  def executedPlan: SparkPlan = currentPhysicalPlan
+
+  override def output: Seq[Attribute] = initialPlan.output
+
+  override def doCanonicalize(): SparkPlan = initialPlan.canonicalized
+
+  override def doExecute(): RDD[InternalRow] = {
+    val events = new LinkedBlockingQueue[StageMaterializationEvent]()
+    var result = createQueryStages(currentPhysicalPlan)
+    while(!result.allChildStagesMaterialized) {
+      currentPhysicalPlan = result.newPlan
+      updateLogicalPlan(result.newStages)
+      onUpdatePlan()
+      result.newStages.map(_._2).foreach { stage =>
+        stage.materialize().onComplete { res =>
+          if (res.isSuccess) {
+            stage.resultOption = Some(res.get)
+            events.offer(StageSuccess(stage))
+          } else {
+            events.offer(StageFailure(stage, res.failed.get))
+          }
+        }
+      }
+      // Wait on the next completed stage, which indicates new stats are available and probably
+      // new stages can be created. There might be other stages that finish at around the same
+      // time, so we process those stages too in order to reduce re-planning.
+      val nextMsg = events.take()
+      val rem = mutable.ArrayBuffer.empty[StageMaterializationEvent]
```

Comment:
This comment got dropped after the recent push. `rem` is not being used. I think you meant to drain all other completed events here.

Reply:
Yes :)